src/main/java/com/vadmack/mongodbtest/service/FileService.java

```diff
@@ -30,66 +32,128 @@ public void init() {
         try {
             Files.createDirectories(Paths.get(rootDirectory));
         } catch (IOException ex) {
-            throw new ServerSideException("Could not create upload folder: " + ex.getMessage());
+            throw new ServerSideException("Could not create upload folder: " + ex.getMessage(), ex);
         }
     }
 
     public void saveFromLink(String url) {
         String[] urlParts = url.split("/");
         String filename = urlParts[urlParts.length - 1];
-        metadataService.checkAlreadyExists(filename);
+        metadataService.checkAlreadyExists("", filename);
 
-        File file = new File(String.valueOf(Paths.get(rootDirectory).resolve(filename)));
-        try (InputStream in = new URL(url).openStream()) {
-            Files.copy(in, file.toPath());
-            metadataService.create(file);
-        } catch (IOException ex) {
-            throw new ServerSideException("Could not save file: " + ex.getMessage());
-        }
+        File file = getAbsolutePathOfFile(filename).toFile();
+        Long metadataId = metadataService.create(file, FileMetadata.Status.NOT_READY_FOR_USE).getId();
+
+        Runnable task = () -> {
+            try (InputStream in = new URL(url).openStream()) {
+                Files.copy(in, file.toPath());
+                FileMetadata updatedMetadata = metadataService
+                        .buildMetadataFromFile(file, FileMetadata.Status.READY_FOR_USE);
+                updatedMetadata.setId(metadataId);
+                metadataService.update(updatedMetadata);
+            } catch (IOException ex) {
+                throw new ServerSideException("Could not save file: " + ex.getMessage(), ex);
+            }
+        };
+
+        Thread fileSaveThread = new Thread(task);
+        fileSaveThread.start();
     }
 
     public Resource load(Long id) {
         FileMetadata metadata = metadataService.findById(id);
-        return new FileSystemResource(Paths.get(rootDirectory).resolve(metadata.getDirectory()));
+        return new FileSystemResource(getAbsolutePathOfFile(metadata));
     }
 
     public void delete(Long id) {
         FileMetadata metadata = metadataService.findById(id);
-        try {
-            Files.deleteIfExists(Paths.get(rootDirectory).resolve(metadata.getDirectory()));
-            metadataService.delete(id);
-        } catch (IOException ex) {
-            throw new ServerSideException("Could not delete file: " + ex.getMessage());
-        }
+        checkFileStatus(metadata);
```

Comment:
It would be more interesting to  interrupt the upload or movement operation (for example) in case if the file becomes useless for a user